File: Cpp/Count_Square_Submatrices_with_All_Ones.cpp
class Solution {
public:
    int countSquares(vector<vector<int>>& nums) {
        vector<vector<int>> dp(nums.size(), vector<int>(nums[0].size(), 0));
        long ans = 0;
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < nums[0].size(); j++) {
                if (nums[i][j] == 1) {
                    if (i == 0 || j == 0) {
                        dp[i][j] = 1;
                    } else {
                        dp[i][j] = 1 + min({dp[i - 1][j - 1], dp[i - 1][j],
                                            dp[i][j - 1]});
                    }
                }
                ans += dp[i][j];
            }
        }
        return ans;
    }
};

File: Rust/Count_Square_Submatrices_with_All_Ones.rs
impl Solution {
    pub fn count_squares(matrix: Vec<Vec<i32>>) -> i32 {
        let n = matrix.len();
        let m = matrix[0].len();

        let mut dp = vec![vec![0; m]; n];
        let mut ans = 0;

        for i in 0..n {
            for j in 0..m {
                if matrix[i][j] == 1 {
                    if i == 0 || j == 0 {
                        dp[i][j] = 1;
                    } else {
                        let min_val = dp[i - 1][j]
                            .min(dp[i][j - 1])
                            .min(dp[i - 1][j - 1]);
                        dp[i][j] = 1 + min_val;
                    }
                }
                ans += dp[i][j];
            }
        }

        ans
    }
}
